tower-buffer/src/lib.rs

```diff
@@ -31,25 +30,36 @@ pub struct Buffer<T, Request>
 where
     T: Service<Request>,
 {
-    tx: mpsc::Sender<Message<Request, T::Future>>,
-    state: Arc<State>,
+    tx: mpsc::Sender<Message<Request, T::Future, T::Error>>,
+    state: Arc<State<T::Error>>,
 }
 
 /// A [`Buffer`] that is backed by a `DirectService`.
 pub type DirectBuffer<T, Request> = Buffer<DirectServiceRef<T>, Request>;
 
 /// Future eventually completed with the response to the original request.
-pub struct ResponseFuture<T> {
-    state: ResponseState<T>,
+pub struct ResponseFuture<T, E> {
+    state: ResponseState<T, E>,
+}
+
+/// An error produced by a `Service` wrapped by a `Buffer`
+#[derive(Debug)]
+pub struct ServiceError<E> {
+    /// The method that was called on `Service` when it failed.
+    pub method: &'static str,
+    /// The error produced by the `Service`.
+    pub inner: E,
 }
 
 /// Errors produced by `Buffer`.
 #[derive(Debug)]
-pub enum Error<T> {
+pub enum Error<E> {
```

Comment:
TIOLI: While we're here, should we maybe change this into a struct for forwards-compatibility reasons?

Reply:
I'm not sure why we'd want this to be a struct? If you're thinking of #131, then we could still expose an enum as a `dyn Error` in the future?

Reply:
@jonhoo definitely can be part of a separate PR.